Modify zero_state to work with MultiRNNCell when state_is_tuple is True

jihunchoi opened the pull request:
#2463

tensorflow-jenkins commented:
Can one of the admins verify this patch?

jihunchoi commented:
@googlebot I signed a CLA.

vrv commented:
Thanks for the contribution, can you add a test for these changes so that we don't regress in the future?

jihunchoi commented:
Implemented a test on my implementation.
Since the current version of `testBasicLSTMCellWithStateTuple` in rnn_cell_test.py is not properly using state_is_tuple, (it is using state_is_tuple only on `MultiRNNCell`, not on `BasicLSTMCel`l), I modified `testBasicLSTMCellWithStateTuple` to use state_is_tuple on both cells.

vrv commented:
Looks good!  Thanks.  @tensorflow-jenkins test this please

vrv commented on tensorflow/models/rnn/ptb/ptb_word_lm.py:
```diff
@@ -245,13 +245,24 @@ def run_epoch(session, m, data, eval_op, verbose=False):
+    feed_dict[m.targets] = y
+    for i, (c, h) in enumerate(m.initial_state):
+        feed_dict[c], feed_dict[h] = state[i]
+    cost, _, *states = session.run(fetches, feed_dict)
```
I think this is an error according to our linter

jihunchoi replied:
My text editor inserted 4 spaces for indent instead of 2 spaces...
Updated the codes, and checked that it passes the sanity checks in my local machine.

jihunchoi replied:
Wait; it only runs on Python 3.
I will modify soon.

jihunchoi replied:
I've just modified the codes to be compatible with Python 2 and checked that it passes sanity checks on my 2.7.11 and 3.5.1.
Can you test via jenkins please?

vrv commented:
Sure, will kick off a test:  @tensorflow-jenkins test this please

@ebrevdo should still look at this though

vrv commented:
Some test failures -- you can test locally by running bazel test tensorflow/... to make sure you catch all failures.

ebrevdo commented:
We have a change internally that should fix nested tuples.  It obsoletes
the rnn changes in this pr.  At our next public push you should rebase and
we can see about the ptb example updates.
On May 24, 2016 9:09 PM, "Vijay Vasudevan" <notifications@github.com> wrote:

> Sure, will kick off a test: @tensorflow-jenkins
> <https://github.com/tensorflow-jenkins> test this please
>
> @ebrevdo <https://github.com/ebrevdo> should still look at this though
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/2490#issuecomment-221468342>
>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

jihunchoi commented:
Wow, the test works properly after completely reinstalling the whole library...

I didn't know that this is internally fixed, so I will left only PTB examples.
